src/library/scala/collection/mutable/ArrayBuffer.scala

```diff
@@ -159,12 +159,14 @@ class ArrayBuffer[A] private (initialElements: Array[AnyRef], initialSize: Int)
 
   override protected[this] def stringPrefix = "ArrayBuffer"
 
-  override def copyToArray[B >: A](xs: Array[B], start: Int): xs.type = copyToArray[B](xs, start, length)
+  override def copyToArray[B >: A](xs: Array[B], start: Int): Int = copyToArray[B](xs, start, length)
 
-  override def copyToArray[B >: A](xs: Array[B], start: Int, len: Int): xs.type = {
-    val l = scala.math.min(scala.math.min(len, length), xs.length-start)
-    if(l > 0) Array.copy(array, 0, xs, start, l)
-    xs
+  override def copyToArray[B >: A](xs: Array[B], start: Int, len: Int): Int = {
+    val _len = scala.math.min(scala.math.min(len, length), xs.length-start)
```

Comment:
Also, I’m thinking that it could be beneficial to define only once the computation that returns the number of copied elements, and then reuse it everywhere so that we don’t miss the final `max(…, 0)`. You could, for instance, create a private method in the `IterableOnce` companion object.

Reply:
We could even put the `if (copied > 0)` in that reusable method to save some logic duplication…

~~~ scala
@inline private[collection] def copyToArrayTemplate[A](
    thisLength: Int, xs: Array[A], start: Int, len: Int)(doCopy: Int => Unit): Int = {
  val copied = math.max(math.min(scala.math.min(len, thisLength), xs.length-start), 0)
  if (copied > 0) {
    doCopy(copied)
  }
  copied
}
~~~

And then, at use site:

~~~ scala
override def copyToArray[B >: A](xs: Array[B], start: Int, len: Int): Int =
  copyToArrayTemplate(length, xs, start, len) { copied =>
    Array.copy(array, 0, xs, start, copied)
  }
~~~

WDYT?

(You can use a better name than `copyToArrayTemplate`, obviously…)

(And I’m not sure how efficiently are literal functions inlined/optimized in such cases)

Reply:
Hey @julienrf I really like the idea of implementing this a single time. I think though that the utility method could just return the number of elements to write, and that would save us having to allocate a clojure